### PR TITLE
WIP : brpc server

### PR DESCRIPTION
**Pull request by pingfengluo**
- [x] brpc server

- [x] support merge network IO for dynamic batch decoding in max_wait_time

### CODE REVIEW

File: sherpa/cpp_api/brpc/client.cc [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+#include "asr.pb.h" //NOLINT
+#include "brpc/channel.h"
+#include "brpc/stream.h"
+#include "bthread/bthread.h"
+#include "butil/logging.h"
+#include "gflags/gflags.h"
+
+
+DEFINE_string(protocol, "baidu_std",
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/client.cc:28]: Can we remove gflags and use `parse_options.{h,cc}` to replace it?

Comment (pingfengluo) [reply sherpa/cpp_api/brpc/client.cc]: ok, done.

File: sherpa/cpp_api/brpc/client.cc [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+  size_t frame_size = sample_rate * frame_time * sizeof(int16_t);
+  std::vector<char> pcm(frame_size);
+  FILE * fp = fopen(wav_file.c_str(), "rb");
+  if (fp) {
+    fseek(fp, 44, SEEK_SET);
+  }
+
+  while (fp &&
+      (frame_size =
+       fread(reinterpret_cast<char*>(pcm.data()), sizeof(char), frame_size, fp))
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/client.cc:123]:
```suggestion
       fread(static_cast<void*>(pcm.data()), sizeof(char), frame_size, fp))
```
or use just ` fread(pcm.data(), sizeof(char), frame_size, fp)`

The type of `pcm.data()` is already `char*`.

File: sherpa/cpp_api/brpc/client.cc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      >= 0) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/client.cc:124]:
```suggestion
      > 0) {
```

Should it be `> 0`. If it is equal to 0, it means we reach the end of the file or an error occurs.

File: sherpa/cpp_api/brpc/client.cc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        frame_size);
+
+    sherpa::AsrRequest_Status status = sherpa::AsrRequest::stream_continue;
+    if (frame_size == 0) {
+      status = sherpa::AsrRequest::stream_end;
+    }
+    request.set_status(status);
+
+    // in asynchronous RPC, we fetch the result inside the callback
+    google::protobuf::Closure* done = brpc::NewCallback(
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/client.cc:145]: When will `done` be freed?

File: sherpa/cpp_api/brpc/client.cc [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+    size_t sleep_us = frame_time * 1000000;
+    if (frame_size == 0) {
+      sleep_us *= 10;
+    }
+    bthread_usleep(sleep_us);
+    if (frame_size == 0) {
+      break;
+    }
+  }
+  while (!brpc::IsAskedToQuit()) {}
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/client.cc:157]: Could you add comments for this`while` loop?

File: sherpa/cpp_api/brpc/server.cc [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+#include "sherpa/csrc/parse_options.h"
+
+#include "brpc/server.h"
+#include "brpc/stream.h"
+#include "butil/base64.h"
+#include "butil/logging.h"
+
+using namespace std::chrono_literals;  //NOLINT
+
+const int32_t MAX_ASR_RESULT = 512;
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:38]: To follow the google code style, we usually use
`kMaxAsrResult` for a const variable.

File: sherpa/cpp_api/brpc/server.cc [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+  input_end,
+};
+// set endpoint_reset state in two case:
+// 1) endpoint inactive and already fetch result
+// 2) no more input
+
+// AsrDecoder
+class AsrDecoder final {
+ public:
+  AsrDecoder(const sherpa::OnlineAsrOptions opts,
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:56]:
```suggestion
  AsrDecoder(const sherpa::OnlineAsrOptions& opts,
```

File: sherpa/cpp_api/brpc/server.cc [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+      const size_t max_batch_size = 4096,
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:57]: Shall we use `int32_t` for `max_batch_size`?

File: sherpa/cpp_api/brpc/server.cc [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+  brpc::Server server;
+  StreamingAsrService asr_service_impl(opts);
+  if (server.AddService(&asr_service_impl,
+        brpc::SERVER_DOESNT_OWN_SERVICE) != 0) {
+    LOG(ERROR) << "Fail to add service";
+    return -1;
+  }
+  brpc::ServerOptions options;
+  options.idle_timeout_sec = idle_timeout_s;
+  if (server.Start(port, &options) != 0) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:372]: Does it by default listen on all the available interfaces of the current machine?

File: sherpa/cpp_api/brpc/server.cc [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    tail_padding_(SAMPLE_RATE * sizeof(int16_t) * PADDING_LEN, 0) {}
+
+  ~StreamingAsrService() {
+    decoder_->stop_decode();
+  }
+
+  virtual void Recognize(google::protobuf::RpcController* controller,
+      const sherpa::AsrRequest * request,
+      sherpa::AsrResponse* response,
+      google::protobuf::Closure* done) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:260]:
```suggestion
  void Recognize(google::protobuf::RpcController* controller,
      const sherpa::AsrRequest * request,
      sherpa::AsrResponse* response,
      google::protobuf::Closure* done) override {
```

File: sherpa/cpp_api/brpc/server.cc [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+    try {
+      sherpa::AsrResponse_Status status = sherpa::AsrResponse::ok;
+      brpc::ClosureGuard done_guard(done);
+      brpc::Controller* cntl = static_cast<brpc::Controller*>(controller);
+      std::string remote_name(endpoint2str(cntl->remote_side()).c_str());
+      if (remote_id_.find(remote_name) == remote_id_.end()) {
+        std::unique_lock<std::mutex> mlock(mutex_);
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:267]:
```suggestion
        std::lock_guard<std::mutex> mlock(mutex_);
```

`std::lock_guard` is more lightweight.

File: sherpa/cpp_api/brpc/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+find_library(THRIFT_LIB NAMES thrift)
+if (NOT THRIFT_LIB)
+    set(THRIFT_LIB "")
+endif()
+find_library(THRIFTNB_LIB NAMES thriftnb)
+if (NOT THRIFTNB_LIB)
+    set(THRIFTNB_LIB "")
+endif()
+
+find_path(GFLAGS_INCLUDE_PATH gflags/gflags.h)
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/CMakeLists.txt:14]: Do we still need glfags?

File: sherpa/cpp_api/brpc/server.cc [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+#include "butil/base64.h"
+#include "butil/logging.h"
+
+using namespace std::chrono_literals;  //NOLINT
+
+const int32_t MAX_ASR_RESULT = 512;
+const int32_t SAMPLE_RATE = 16000;
+const float PADDING_LEN = 0.36;
+enum StreamState : int32_t {
+  unknow_state = -1,
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:42]: For enum names, we usually use
```cpp
kUnknownState
```

File: sherpa/cpp_api/brpc/server.cc [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+
+  ~AsrDecoder() {
+    while (!exit_) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+  }
+
+  int64_t acquire_stream() {
+    int64_t stream_id = -1;
+    std::shared_lock mlock(mutex_);
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:81]: Is there a reason to use `shared_lock` here?
I think we are modifying `stream_queue` inside, so `std::unique_lock` maybe is more appropriate here.

File: sherpa/cpp_api/brpc/server.cc [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+      SHERPA_LOG(WARNING) << "stream_queue_.size() and max_batch_size_ = "
+        << max_batch_size_ << ", can't create more stream";
+    }
+    return stream_id;
+  }
+
+  void release_stream(const int64_t stream_id) {
+    std::shared_lock mlock(mutex_);
+    if (stream_queue_.find(stream_id) != stream_queue_.end()) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:102]: Is there a reason to sleep here?

File: sherpa/cpp_api/brpc/server.cc [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+      batch_cond_.notify_one();
+      if (vad_state_queue_[stream_id] != endpoint_reset) {
+        if (segment_stream(stream_id)) {
+          vad_state_queue_[stream_id] = endpoint_active;
+        }
+      }
+    }
+  }
+
+  const char * fetch_stream(const int64_t stream_id) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:158]: Since it requires c++17, I suggest that you use a return value of type `std::optional<std::string>`.

File: sherpa/cpp_api/brpc/server.cc [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+        decoding_ = false;
+      }
+      decoding_cond_.notify_all();
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    exit_ = true;
+  }
+
+  void stop_decode() {
+    std::unique_lock mlock(mutex_);
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:214]:
```suggestion
    std::lock_guard mlock(mutex_);
```

If you need to transfer the ownership, then you can use `unique_xxx`, e.g., `std::unique_ptr`, `std::unique_lock`.

File: sherpa/cpp_api/brpc/server.cc [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+  }
+
+  void warmup_decode() {
+    auto s = online_asr_.CreateStream();
+    auto wav_tensor = torch::rand({32000}).to(torch::kFloat);
+    s->AcceptWaveform(SAMPLE_RATE, wav_tensor);
+  }
+
+ private:
+  mutable std::shared_mutex mutex_;
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:225]:
```suggestion
 std::shared_mutex mutex_;
```

File: sherpa/cpp_api/brpc/server.cc [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+      batch_size_ = std::max(1,
+          std::min(static_cast<int32_t>(max_batch_size_ >> 2), 64));
+      decoder_ = std::thread(&AsrDecoder::batch_decode, this);
+      decoder_.detach();
+      srand(42);
+      warmup_decode();
+    }
+
+  ~AsrDecoder() {
+    while (!exit_) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:74]: Instead of using polling, I suggest that we can use
`std::future<void>` and `std::promise<void>` here.
You can set the promise in the `decoder_` thread.

----

Alternatively, we don't call `decoder_.detach()` in the constructor but use `decoder_.join()` here in the destructor.

File: sherpa/cpp_api/brpc/server.cc [hunk 20]
```diff
@@ -0,0 +1,386 @@
...
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+  }
+
+  int64_t acquire_stream() {
+    int64_t stream_id = -1;
```
Comment (csukuangfj) [inline sherpa/cpp_api/brpc/server.cc:80]: Can we use `std::atomic<int64_t>` or `std::atomic<int32_t>` to generate a stream ID that is monotonically increasing?

**Comment by pingfengluo** [comment]
Sherpa already has  a C++ server, close brpc now.
